src/Tes.Runner/Transfer/ProcessedPartsProcessor.cs

```diff
@@ -62,15 +69,25 @@ public ProcessedPartsProcessor(IBlobPipeline blobPipeline)
                 }
             }
 
-        await PartsProcessor.WhenAllOrThrowIfOneFailsAsync(tasks);
-
-        readBufferChannel.Writer.Complete();
+        try
+        {
+            await PartsProcessor.WhenAllOrThrowIfOneFailsAsync(tasks);
+        }
+        finally
+        {
+            readBufferChannel.Writer.Complete();
+        }
 
         logger.LogInformation("All parts were successfully processed.");
 
         return totalBytes;
     }
 
+    private string GetRootHashIfProviderIsSet(IHashListProvider hashListProvider)
```

Comment:
This name is misleading. It communicates that if the provider is not set (aka is null) it will not get the root hash. Instead, it would throw an exception because there's no conditional logic here.

Reply:
I agree and will change it.